File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Cdm/CdmTraitDefinition.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Cdm
{
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.ResolvedModel;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using System;
    using System.Collections.Generic;

    public class CdmTraitDefinition : CdmObjectDefinitionBase
    {
        /// <summary>
        /// Gets or sets the trait name.
        /// </summary>
        public string TraitName { get; set; }

        /// <inheritdoc />
        public override string GetName()
        {
            return this.TraitName;
        }

        /// <summary>
        /// Gets or sets the trait extended by this trait.
        /// </summary>
        public CdmTraitReference ExtendsTrait { get; set; }

        private CdmCollection<CdmParameterDefinition> _hasParameters { get; set; }

        ParameterCollection AllParameters { get; set; }

        private bool HasSetFlags;

        /// <summary>
        /// Gets or sets whether the trait is elevated.
        /// </summary>
        public bool? Elevated { get; set; }

        /// <summary>
        /// Gets or sets whether trait is user facing or not.
        /// </summary>
        public bool? Ugly { get; set; }

        /// <summary>
        /// Gets or sets the trait associated properties.
        /// </summary>
        public List<string> AssociatedProperties { get; set; }

        bool? BaseIsKnownToHaveParameters { get; set; }

        internal bool? ThisIsKnownToHaveParameters { get; set; }

        /// <summary>
        /// Gets the trait parameters.
        /// </summary>
        public CdmCollection<CdmParameterDefinition> Parameters
        {
            get
            {
                if (this._hasParameters == null)
                    this._hasParameters = new CdmCollection<CdmParameterDefinition>(this.Ctx, this, CdmObjectType.ParameterDef);
                return this._hasParameters;
            }
        }

        /// <summary>
        /// Constructs a CdmTraitDefinition.
        /// </summary>
        /// <param name="ctx">The context.</param>
        /// <param name="name">The trait name.</param>
        /// <param name="extendsTrait">The trait extended by this trait.</param>
        public CdmTraitDefinition(CdmCorpusContext ctx, string name, CdmTraitReference extendsTrait = null)
            : base(ctx)
        {
            this.HasSetFlags = false;
            this.ObjectType = CdmObjectType.TraitDef;
            this.TraitName = name;
            this.ExtendsTrait = extendsTrait;
        }

        /// <inheritdoc />
        public override CdmObject Copy(ResolveOptions resOpt = null, CdmObject host = null)
        {
            if (resOpt == null)
            {
                resOpt = new ResolveOptions(this, this.Ctx.Corpus.DefaultResolutionDirectives);
            }

            CdmTraitDefinition copy;
            if (host == null)
            {
                copy = new CdmTraitDefinition(this.Ctx, this.TraitName, null);
            }
            else
            {
                copy = host as CdmTraitDefinition;
                copy.Ctx = this.Ctx;
                copy.TraitName = this.TraitName;
            }

            copy.ExtendsTrait = (CdmTraitReference)this.ExtendsTrait?.Copy(resOpt);
            copy.AllParameters = null;
            copy.Elevated = this.Elevated;
            copy.Ugly = this.Ugly;
            copy.AssociatedProperties = this.AssociatedProperties;

            this.CopyDef(resOpt, copy);
            return copy;
        }

        [Obsolete("CopyData is deprecated. Please use the Persistence Layer instead.")]
        public override dynamic CopyData(ResolveOptions resOpt, CopyOptions options)
        {
            return CdmObjectBase.CopyData<CdmTraitDefinition>(this, resOpt, options);
        }

        internal ParameterCollection FetchAllParameters(ResolveOptions resOpt)
        {
            if (this.AllParameters != null)
                return this.AllParameters;

            // get parameters from base if there is one
            ParameterCollection prior = null;
            if (this.ExtendsTrait != null)
                prior = this.ExtendsTrait.FetchObjectDefinition<CdmTraitDefinition>(resOpt).FetchAllParameters(resOpt);
            this.AllParameters = new ParameterCollection(prior);
            if (this.Parameters != null)
            {
                foreach (CdmParameterDefinition element in this.Parameters)
                {
                    this.AllParameters.Add(element as CdmParameterDefinition);
                }
            }
            return this.AllParameters;
        }

        [Obsolete]
        public override CdmObjectType GetObjectType()
        {
            return CdmObjectType.TraitDef;
        }

        /// <inheritdoc />
        public override bool IsDerivedFrom(string baseDef, ResolveOptions resOpt = null)
        {
            if (resOpt == null)
            {
                resOpt = new ResolveOptions(this, this.Ctx.Corpus.DefaultResolutionDirectives);
            }

            if (baseDef == this.TraitName)
                return true;
            return this.IsDerivedFromDef(resOpt, this.ExtendsTrait, this.TraitName, baseDef);
        }

        /// <inheritdoc />
        public override bool Validate()
        {
            if (string.IsNullOrWhiteSpace(this.TraitName))
            {
                Logger.Error(nameof(CdmTraitDefinition), this.Ctx, Errors.ValidateErrorString(this.AtCorpusPath, new List<string> { "TraitName" }), nameof(Validate));
                return false;
            }
            return true;
        }

        /// <inheritdoc />
        public override bool Visit(string pathFrom, VisitCallback preChildren, VisitCallback postChildren)
        {
            string path = string.Empty;
            if (this.Ctx.Corpus.blockDeclaredPathChanges == false)
            {
                path = this.DeclaredPath;
                if (string.IsNullOrEmpty(path))
                {
                    path = pathFrom + this.TraitName;
                    this.DeclaredPath = path;
                }
            }

            if (preChildren != null && preChildren.Invoke(this, path))
                return false;
            if (this.ExtendsTrait != null)
            {
                this.ExtendsTrait.Owner = this;
                if (this.ExtendsTrait.Visit(path + "/extendsTrait/", preChildren, postChildren))
                    return true;
            }
            if (this.Parameters != null)
                if (this.Parameters.VisitList(path + "/hasParameters/", preChildren, postChildren))
                    return true;
            if (postChildren != null && postChildren.Invoke(this, path))
                return true;
            return false;
        }

        internal override ResolvedTraitSet FetchResolvedTraits(ResolveOptions resOpt = null)
        {
            if (resOpt == null)
            {
                resOpt = new ResolveOptions(this, this.Ctx.Corpus.DefaultResolutionDirectives);
            }

            const string kind = "rtsb";
            ResolveContext ctx = this.Ctx as ResolveContext;
            // this may happen 0, 1 or 2 times. so make it fast
            CdmTraitDefinition baseTrait = null;
            ResolvedTraitSet baseRts = null;
            List<dynamic> baseValues = null;
            System.Action GetBaseInfo = () =>
            {
                if (this.ExtendsTrait != null)
                {
                    baseTrait = this.ExtendsTrait.FetchObjectDefinition<CdmTraitDefinition>(resOpt);
                    if (baseTrait != null)
                    {
                        baseRts = this.ExtendsTrait.FetchResolvedTraits(resOpt);
                        if (baseRts?.Size == 1)
                        {
                            ParameterValueSet basePv = baseRts.Get(baseTrait)?.ParameterValues;
                            if (basePv != null)
                                baseValues = basePv.Values;
                        }
                    }
                }
            };

            // see if one is already cached
            // if this trait has parameters, then the base trait found through the reference might be a different reference
            // because trait references are unique per argument value set. so use the base as a part of the cache tag
            // since it is expensive to figure out the extra tag, cache that too!
            if (this.BaseIsKnownToHaveParameters == null)
            {
                GetBaseInfo();
                // is a cache tag needed? then make one
                this.BaseIsKnownToHaveParameters = false;
                if (baseValues?.Count > 0)
                    this.BaseIsKnownToHaveParameters = true;
            }
            string cacheTagExtra = "";
            if (this.BaseIsKnownToHaveParameters == true)
                cacheTagExtra = this.ExtendsTrait.Id.ToString();

            string cacheTag = ctx.Corpus.CreateDefinitionCacheTag(resOpt, this, kind, cacheTagExtra);
            dynamic rtsResultDynamic = null;
            if (cacheTag != null)
                ctx.Cache.TryGetValue(cacheTag, out rtsResultDynamic);
            ResolvedTraitSet rtsResult = rtsResultDynamic as ResolvedTraitSet;

            // store the previous reference symbol set, we will need to add it with
            // children found from the constructResolvedTraits call 
            SymbolSet currSymbolRefSet = resOpt.SymbolRefSet;
            if (currSymbolRefSet == null)
                currSymbolRefSet = new SymbolSet();
            resOpt.SymbolRefSet = new SymbolSet();

            // if not, then make one and save it
            if (rtsResult == null)
            {
                GetBaseInfo();
                if (baseTrait != null)
                {
                    // get the resolution of the base class and use the values as a starting point for this trait's values
                    if (!this.HasSetFlags)
                    {
                        // inherit these flags
                        if (this.Elevated == null)
                            this.Elevated = baseTrait.Elevated;
                        if (this.Ugly == null)
                            this.Ugly = baseTrait.Ugly;
                        if (this.AssociatedProperties == null)
                            this.AssociatedProperties = baseTrait.AssociatedProperties;
                    }
                }
                this.HasSetFlags = true;
                ParameterCollection pc = this.FetchAllParameters(resOpt);
                List<dynamic> av = new List<dynamic>();
                List<bool> wasSet = new List<bool>();
                this.ThisIsKnownToHaveParameters = pc.Sequence.Count > 0;
                for (int i = 0; i < pc.Sequence.Count; i++)
                {
                    // either use the default value or (higher precidence) the value taken from the base reference
                    dynamic value = (pc.Sequence[i] as CdmParameterDefinition).DefaultValue;
                    dynamic baseValue = null;
                    if (baseValues != null && i < baseValues.Count)
                    {
                        baseValue = baseValues[i];
                        if (baseValue != null)
                            value = baseValue;
                    }
                    av.Add(value);
                    wasSet.Add(false);
                }

                // save it
                ResolvedTrait resTrait = new ResolvedTrait(this, pc, av, wasSet);
                rtsResult = new ResolvedTraitSet(resOpt);
                rtsResult.Merge(resTrait, false);

                // register set of possible symbols
                ctx.Corpus.RegisterDefinitionReferenceSymbols(this.FetchObjectDefinition<CdmObjectDefinitionBase>(resOpt), kind, resOpt.SymbolRefSet);
                // get the new cache tag now that we have the list of docs
                cacheTag = ctx.Corpus.CreateDefinitionCacheTag(resOpt, this, kind, cacheTagExtra);
                if (!string.IsNullOrWhiteSpace(cacheTag))
                    ctx.Cache[cacheTag] = rtsResult;
            }
            else
            {
                // cache found
                // get the SymbolSet for this cached object
                string key = CdmCorpusDefinition.CreateCacheKeyFromObject(this, kind);
                ((CdmCorpusDefinition)ctx.Corpus).DefinitionReferenceSymbols.TryGetValue(key, out SymbolSet tempSymbolRefSet);
                resOpt.SymbolRefSet = tempSymbolRefSet;
            }
            // merge child document set with current
            currSymbolRefSet.Merge(resOpt.SymbolRefSet);
            resOpt.SymbolRefSet = currSymbolRefSet;

            return rtsResult;
        }

        internal override void ConstructResolvedTraits(ResolvedTraitSetBuilder rtsb, ResolveOptions resOpt)
        {
            return;
        }

        internal override ResolvedAttributeSetBuilder ConstructResolvedAttributes(ResolveOptions resOpt, CdmAttributeContext under = null)
        {
            return null;
        }
    }
}

File: objectModel/TypeScript/Cdm/CdmTraitDefinition.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import {
    ArgumentValue,
    CdmAttributeContext,
    CdmCollection,
    CdmCorpusContext,
    CdmCorpusDefinition,
    CdmObject,
    CdmObjectDefinitionBase,
    cdmObjectType,
    CdmParameterDefinition,
    CdmTraitReference,
    cdmLogCode,
    Logger,
    ParameterCollection,
    ParameterValueSet,
    resolveContext,
    ResolvedAttributeSetBuilder,
    ResolvedTrait,
    ResolvedTraitSet,
    ResolvedTraitSetBuilder,
    resolveOptions,
    SymbolSet,
    VisitCallback
} from '../internal';

export class CdmTraitDefinition extends CdmObjectDefinitionBase {
    private TAG: string = CdmTraitDefinition.name;

    public static get objectType(): cdmObjectType {
        return cdmObjectType.traitDef;
    }

    public get parameters(): CdmCollection<CdmParameterDefinition> {
        if (!this._parameters) {
            this._parameters = new CdmCollection<CdmParameterDefinition>(this.ctx, this, cdmObjectType.parameterDef);
        }

        return this._parameters;
    }
    public associatedProperties: string[];
    public explanation: string;
    public elevated: boolean;
    public traitName: string;
    public extendsTrait?: CdmTraitReference;
    public ugly: boolean;
    /**
     * @internal
     */
    public thisIsKnownToHaveParameters: boolean;

    private _parameters: CdmCollection<CdmParameterDefinition>;
    private allParameters: ParameterCollection;
    private hasSetFlags: boolean;
    private baseIsKnownToHaveParameters: boolean;

    constructor(ctx: CdmCorpusContext, name: string, extendsTrait?: CdmTraitReference) {
        super(ctx);
        // let bodyCode = () =>
        {
            this.hasSetFlags = false;
            this.objectType = cdmObjectType.traitDef;
            this.traitName = name;
            this.extendsTrait = extendsTrait;
        }
        // return p.measure(bodyCode);
    }

    public getObjectType(): cdmObjectType {
        // let bodyCode = () =>
        {
            return cdmObjectType.traitDef;
        }
        // return p.measure(bodyCode);
    }

    public copy(resOpt?: resolveOptions, host?: CdmObject): CdmObject {
        // let bodyCode = () =>
        {
            if (!resOpt) {
                resOpt = new resolveOptions(this, this.ctx.corpus.defaultResolutionDirectives);
            }
            let copy: CdmTraitDefinition;
            if (!host) {
                copy = new CdmTraitDefinition(this.ctx, this.traitName, undefined);
            } else {
                copy = host as CdmTraitDefinition;
                copy.ctx = this.ctx;
                copy.traitName = this.traitName;
            }
            copy.extendsTrait = this.extendsTrait
                ? <CdmTraitReference>this.extendsTrait.copy(resOpt)
                : undefined;
            copy.allParameters = undefined;
            copy.elevated = this.elevated;
            copy.ugly = this.ugly;
            copy.associatedProperties = this.associatedProperties;
            this.copyDef(resOpt, copy);

            return copy;
        }
        // return p.measure(bodyCode);
    }

    public validate(): boolean {
        // let bodyCode = () =>
        {
            if (!this.traitName) {
                let missingFields: string[] = ['traitName'];
                Logger.error(this.ctx, this.TAG, this.validate.name, this.atCorpusPath, cdmLogCode.ErrValdnIntegrityCheckFailure, missingFields.map((s: string) => `'${s}'`).join(', '), this.atCorpusPath);
                return false;
            }

            return true;
        }
        // return p.measure(bodyCode);
    }

    public getName(): string {
        // let bodyCode = () =>
        {
            return this.traitName;
        }
        // return p.measure(bodyCode);
    }

    /**
     * @internal
     */
    public fetchExtendsTrait(): CdmTraitReference {
        // let bodyCode = () =>
        {
            return this.extendsTrait;
        }
        // return p.measure(bodyCode);
    }

    /**
     * @internal
     */
    public hasParameterDefs(): CdmCollection<CdmParameterDefinition> {
        // let bodyCode = () =>
        {
            return this.parameters;
        }
        // return p.measure(bodyCode);
    }

    public isDerivedFrom(base: string, resOpt?: resolveOptions): boolean {
        // let bodyCode = () =>
        {
            if (!resOpt) {
                resOpt = new resolveOptions(this, this.ctx.corpus.defaultResolutionDirectives);
            }

            if (base === this.traitName) {
                return true;
            }

            return this.isDerivedFromDef(resOpt, this.extendsTrait, this.traitName, base);
        }
        // return p.measure(bodyCode);
    }

    public visit(pathFrom: string, preChildren: VisitCallback, postChildren: VisitCallback): boolean {
        // let bodyCode = () =>
        {
            let path: string = '';
            if (!this.ctx.corpus.blockDeclaredPathChanges) {
                path = this.declaredPath;
                if (!path) {
                    path = pathFrom + this.traitName;
                    this.declaredPath = path;
                }
            }

            if (preChildren && preChildren(this, path)) {
                return false;
            }
            if (this.extendsTrait) {
                this.extendsTrait.owner = this;
                if (this.extendsTrait.visit(`${path}/extendsTrait/`, preChildren, postChildren)) {
                    return true;
                }
            }
            if (this.parameters) {
                if (this.parameters.visitArray(`${path}/hasParameters/`, preChildren, postChildren)) {
                    return true;
                }
            }
            if (postChildren && postChildren(this, path)) {
                return true;
            }

            return false;
        }
        // return p.measure(bodyCode);
    }

    /**
     * @internal
     */
    public fetchResolvedTraits(resOpt?: resolveOptions): ResolvedTraitSet {
        // let bodyCode = () =>
        {
            if (!resOpt) {
                resOpt = new resolveOptions(this, this.ctx.corpus.defaultResolutionDirectives);
            }
            const kind: string = 'rtsb';
            const ctx: resolveContext = this.ctx as resolveContext;
            // this may happen 0, 1 or 2 times. so make it fast
            let baseTrait: CdmTraitDefinition;
            let baseRts: ResolvedTraitSet;
            let baseValues: ArgumentValue[];
            const getBaseInfo: () => void
                = (): void => {
                    if (this.extendsTrait) {
                        baseTrait = this.extendsTrait.fetchObjectDefinition<CdmTraitDefinition>(resOpt);
                        if (baseTrait) {
                            baseRts = this.extendsTrait.fetchResolvedTraits(resOpt);
                            if (baseRts && baseRts.size === 1) {
                                const basePv: ParameterValueSet =
                                    baseRts.get(baseTrait) ? baseRts.get(baseTrait).parameterValues : undefined;
                                if (basePv) {
                                    baseValues = basePv.values;
                                }
                            }
                        }
                    }
                };

            // see if one is already cached
            // if this trait has parameters, then the base trait found through the reference might be a different reference
            // because trait references are unique per argument value set. so use the base as a part of the cache tag
            // since it is expensive to figure out the extra tag, cache that too!
            if (this.baseIsKnownToHaveParameters === undefined) {
                getBaseInfo();
                // is a cache tag needed? then make one
                this.baseIsKnownToHaveParameters = false;
                if (baseValues && baseValues.length > 0) {
                    this.baseIsKnownToHaveParameters = true;
                }
            }
            let cacheTagExtra: string = '';
            if (this.baseIsKnownToHaveParameters) {
                cacheTagExtra = this.extendsTrait.ID.toString();
            }

            let cacheTag: string = ctx.corpus.createDefinitionCacheTag(resOpt, this, kind, cacheTagExtra);
            let rtsResult: ResolvedTraitSet = cacheTag ? ctx.cache.get(cacheTag) : undefined;

            // store the previous reference symbol set, we will need to add it with
            // children found from the constructResolvedTraits call
            const currSymRefSet: SymbolSet = resOpt.symbolRefSet || new SymbolSet();
            resOpt.symbolRefSet = new SymbolSet();

            // if not, then make one and save it
            if (!rtsResult) {
                getBaseInfo();
                if (baseTrait) {
                    // get the resolution of the base class and use the values as a starting point for this trait's values
                    if (this.hasSetFlags === false) {
                        // inherit these flags
                        if (this.elevated === undefined) {
                            this.elevated = baseTrait.elevated;
                        }
                        if (this.ugly === undefined) {
                            this.ugly = baseTrait.ugly;
                        }
                        if (this.associatedProperties === undefined) {
                            this.associatedProperties = baseTrait.associatedProperties;
                        }
                    }
                }
                this.hasSetFlags = true;
                const pc: ParameterCollection = this.fetchAllParameters(resOpt);
                const av: (ArgumentValue)[] = [];
                const wasSet: (boolean)[] = [];
                this.thisIsKnownToHaveParameters = (pc.sequence.length > 0);
                for (let i: number = 0; i < pc.sequence.length; i++) {
                    // either use the default value or (higher precidence) the value taken from the base reference
                    let value: ArgumentValue = pc.sequence[i].defaultValue;
                    let baseValue: ArgumentValue;
                    if (baseValues && i < baseValues.length) {
                        baseValue = baseValues[i];
                        if (baseValue) {
                            value = baseValue;
                        }
                    }
                    av.push(value);
                    wasSet.push(false);
                }

                // save it
                const resTrait: ResolvedTrait = new ResolvedTrait(this, pc, av, wasSet);
                rtsResult = new ResolvedTraitSet(resOpt);
                rtsResult.merge(resTrait, false);

                // register set of possible symbols
                ctx.corpus.registerDefinitionReferenceSymbols(this.fetchObjectDefinition(resOpt), kind, resOpt.symbolRefSet);
                // get the new cache tag now that we have the list of docs
                cacheTag = ctx.corpus.createDefinitionCacheTag(resOpt, this, kind, cacheTagExtra);
                if (cacheTag) {
                    ctx.cache.set(cacheTag, rtsResult);
                }
            } else {
                // cache found
                // get the SymbolSet for this cached object
                const key: string = CdmCorpusDefinition.createCacheKeyFromObject(this, kind);
                resOpt.symbolRefSet = ctx.corpus.definitionReferenceSymbols.get(key);
            }
            // merge child document set with current
            currSymRefSet.merge(resOpt.symbolRefSet);
            resOpt.symbolRefSet = currSymRefSet;

            return rtsResult;
        }
        // return p.measure(bodyCode);
    }

    /**
     * @internal
     */
    public constructResolvedTraits(rtsb: ResolvedTraitSetBuilder, resOpt: resolveOptions): void {
        // traits don't have traits.
    }

    /**
     * @internal
     */
    public fetchAllParameters(resOpt: resolveOptions): ParameterCollection {
        // let bodyCode = () =>
        {
            if (this.allParameters) {
                return this.allParameters;
            }

            // get parameters from base if there is one
            let prior: ParameterCollection;
            if (this.extendsTrait) {
                prior = this.fetchExtendsTrait()
                    .fetchObjectDefinition<CdmTraitDefinition>(resOpt)
                    .fetchAllParameters(resOpt);
            }
            this.allParameters = new ParameterCollection(prior);
            if (this.parameters) {
                this.parameters.allItems.forEach((element: CdmParameterDefinition) => {
                    this.allParameters.add(element);
                });
            }

            return this.allParameters;
        }
        // return p.measure(bodyCode);
    }

    /**
     * @internal
     */
    public constructResolvedAttributes(resOpt: resolveOptions, under?: CdmAttributeContext): ResolvedAttributeSetBuilder {
        // let bodyCode = () =>
        {
            return undefined;
        }
        // return p.measure(bodyCode);
    }
}
